cmd/genbindings/clangast.go

```diff
@@ -0,0 +1,341 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+func isSourceLoc(field string) bool {
+	// "file" (and "line" which we don't care about) takes on the last observed
+	// value for certain node kinds - to simplify processing, we expand it in the
+	// parsed AST.
+	// https://github.com/dtolnay/clang-ast?tab=readme-ov-file#source-locations
+	// https://github.com/llvm/llvm-project/blob/26d9cb17a6e655993c991b66b21d5c378256d79b/clang/lib/AST/JSONNodeDumper.cpp#L285
+	switch field {
+	case "loc", "spellingLoc", "expansionLoc", "begin", "end", "attrLoc":
+		return true
+	}
+
+	return false
+}
+
+// HexUint64 is a uint64 that marshals to/from JSON as a hex string.
+type HexUint64 uint64
+
+func (h HexUint64) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf("\"0x%x\"", uint64(h))), nil
+}
+
+func (h *HexUint64) UnmarshalJSON(data []byte) error {
+	var s string
+	if err := json.Unmarshal(data, &s); err != nil {
+		return err
+	}
+	val, err := strconv.ParseUint(strings.TrimPrefix(s, "0x"), 16, 64)
+	if err != nil {
+		return err
+	}
+	*h = HexUint64(val)
+	return nil
+}
+
+func (h HexUint64) String() string {
+	return fmt.Sprintf("0x%x", uint64(h))
+}
+
+type AstNode struct {
+	Id    HexUint64  `json:"id,omitempty"`
+	Kind  string     `json:"kind"`
+	Inner []*AstNode `json:"inner,omitempty"`
+	// TODO is there a simple way to "flatten" the fields so they appear as in
+	// the original AST?
+	Fields map[string]interface{} `json:"fields,omitempty"`
```

Comment:
I think the only way is to type out the struct definition in full.

It has the downside that sizeof(AstNode) consumes space for empty values for your other types, so ideally you want polymorphic JSON decode into a lot of different AstNode specialized types, which is somewhat annoying to achieve with encoding/json alone.

Reply:
yeah, I had that thought .. miqt only uses a small subset of fields so the overhead is not necessarily that big compared to the instantiation of a map for every node and it would be more convenient to "use" at the expense of making clangast miqt-specific. maybe that's the better tradeoff (vs keeping this module generic / non-miqt-specific)